shared/extension/behaviour/commands.hpp

```diff
@@ -0,0 +1,197 @@
+                                                                     std::function<void(const int&)>))
+            : reaction(reaction_), type(type_), validator(validator_), current(current_), binder(binder_) {}
+
+        /// The Provider reaction that is contingent on this when condition
+        std::shared_ptr<NUClear::threading::Reaction> reaction;
+        /// The enum type that this when expression is looking at
+        std::type_index type;
+        /// Function to determine if the passed state is valid
+        bool (*validator)(const int&);
+        /// Function to get the current state from the global cache
+        int (*current)();
+        /// Function to bind a reaction to monitor when this state changes
+        NUClear::threading::ReactionHandle (*binder)(NUClear::Reactor&, std::function<void(const int&)>);
+    };
+
+    /**
+     * Message used to tell the Director about a new Causing relationship for this Provider
+     */
+    struct CausingExpression {
+
+        /**
+         * Construct a new Causing Expression object to send to the Director
+         *
+         * @param reaction          the reaction this causing condition is for
+         * @param type              the enum type that this causing condition is going to manipulate
+         * @param resulting_state   the state the enum will be in once the causing has succeeded
+         */
+        CausingExpression(const std::shared_ptr<NUClear::threading::Reaction>& reaction,
+                          const std::type_index& type,
+                          const int& resulting_state)
+            : reaction(reaction), type(type), resulting_state(resulting_state) {}
+        /// The Provider reaction that will cause this state
+        std::shared_ptr<NUClear::threading::Reaction> reaction;
+        /// The enum type that this `When` expression claims to manipulate
+        std::type_index type;
+        /// The resulting state this Provider is claiming to provide if it is executed
+        int resulting_state;
+    };
+
+    /**
+     * Message to tell the Director that a Provider task has finished executing
+     */
+    struct ProviderDone {
+
+        /**
+         * @brief Construct a new Provider Done object
+         *
+         * @param requester_id      the reaction_id of the NUClear reaction that finished executing
+         * @param requester_task_id the reaction_task_id of the NUClear reaction task that finished executing
+         */
+        ProviderDone(const uint64_t& requester_id, const uint64_t& requester_task_id)
+            : requester_id(requester_id), requester_task_id(requester_task_id) {}
+        /// The reaction_id of the Provider that finished
+        uint64_t requester_id;
+        /// The specific task_id of the Provider that finished
+        uint64_t requester_task_id;
+    };
+
+    /**
+     * A task message that is to be sent to the Director for execution
+     */
+    struct DirectorTask {
```

Comment:
```suggestion
    template <typename T>
    struct DirectorTask {
```

Reply:
And if I were to do this, how would I later on<Trigger<DirectorTask>>?